juju/juju
# Reject k8s charm upgrades that cannot be deployed

### Checklist


Will need updating when we fully support k8s charms.

----

## Description of change

Current k8s limitations mean that we cannot upgrade a charm to alter the storage requirements because we cannot patch statefulset volume claim templates. The same restriction applies to devices and deployment type info. So we add checks in the api facades to prevent an incompatible charm from being upgraded, otherwise thinsg break in a non obvious way.

To allow for future support of such upgrades (when k8s is fixed), we trigger the k8s deployment worker whenever a charm upgrade hook is flushed, regardless of whether a new pod spec has been sent in. This allows for charm metadata changes to trigger upgrade processing. This is achieved by adding an UpgradeCounter field to the mongo doc holding the podspec. This is always updated even if the podspec has not changed, triggering a notification.

A now unused SetPodSpec() client API was removed from the caas operator facade.

## QA steps

Deploy a k8s charm.
Edit the metadata.yaml to add new storage.
Attempt to upgrade the charm.
See an error saying changing storage is not supported.
